Lib/test/test_tabnanny.py

```diff
@@ -0,0 +1,348 @@
+"""Testing `tabnanny` module.
+
+
+Glossary:
+    * errored    : Whitespace related problems present in file.
+"""
+from unittest import TestCase, mock
+from unittest.mock import patch
+import tabnanny
+import tokenize
+from tempfile import NamedTemporaryFile, TemporaryDirectory
+from test.support import (captured_stderr, captured_stdout, script_helper,
+                          findfile)
+
+
+SOURCE_CODES = {
+    "incomplete_expression": (
+        'fruits = [\n'
+        '    "Apple",\n'
+        '    "Orange",\n'
+        '    "Banana",\n'
+        '\n'
+        'print(fruits)\n'
+    ),
+    "wrong_indented": (
+        'if True:\n'
+        '    print("hello")\n'
+        '  print("world")\n'
+        'else:\n'
+        '    print("else called")\n'
+    ),
+    "nannynag_errored": (
+        'if True:\n'
+        ' \tprint("hello")\n'
+        '\tprint("world")\n'
+        'else:\n'
+        '    print("else called")\n'
+    ),
+    "error_free": (
+        'if True:\n'
+        '    print("hello")\n'
+        '    print("world")\n'
+        'else:\n'
+        '    print("else called")\n'
+    ),
+    "tab_space_errored_1": (
+        'def my_func():\n'
+        '\t  print("hello world")\n'
+        '\t  if True:\n'
+        '\t\tprint("If called")'
+    ),
+    "tab_space_errored_2": (
+        'def my_func():\n'
+        '\t\tprint("Hello world")\n'
+        '\t\tif True:\n'
+        '\t        print("If called")'
+    )
+}
+
+
+def create_tmp_python_file(source_code, dir=None):
+    """Common utility function for writing source code to temporary file.
+
+    It creates file with `.py` extension.
+
+    Returns: Absolute path of created code file."""
+    with NamedTemporaryFile(mode='w', dir=dir,
+                            delete=False, suffix=".py") as f:
```

Comment:
If you only care of creating a temporary filename, use tempfile.mktemp() and then open() manually the file to create it. You might use open(tmpname, "x") to prevent race conditions.